### DISCUSSION TITLE
Long-term access to controllers' logs

### DISCUSSION THREAD

**Discussion by nicogl** [fluxcd/flux2#963] [Q&A] [answered]
Hi,

I deployed a Kubernetes cluster couple of days ago and started using the typical kubectl logs command in order to access the controllers' logs: `kubectl logs -n flux-system -f kustomize-controller-895dfd98d-cr8x7`

It was fine during the first day, but then I noticed that it takes too much time to load and show all the log history. I cannot imagine in the case you have run the cluster for an year period or more... 
Even if I'm interested in the most recent part of the log history, I don't think there's a different way to obtain it, right?

Regardless of that, is there any standard method to record the Flux controllers' log history outside of the cluster, so it's easier to process and manage? I read about Grafana and Prometheus in the [Monitoring ](https://toolkit.fluxcd.io/guides/monitoring/) section of the documentation but I don't think any of them help with the problem I'm describing.

Thanks!

**Comment by phillebaba** [comment, marked as answer]
I would suggest that you look into Elasticsearch or Loki.
https://grafana.com/oss/loki/
https://www.elastic.co/

This is more a generic Kubernetes log exporting question rather than a Flux related question though.

**Comment by fgrelaud** [comment]
Regarding elasticsearch, fluentd-elasticsearch is yout friend (https://github.com/kokuwaio/helm-charts/tree/main/charts/fluentd-elasticsearch)

**Reply by nicogl** [reply]
@phillebaba and @fgrelaud thanks for your answers.
I agree that this is a more generic Kubernetes question, but I raised it in the particular context of Flux Controllers because in order to persist and manage their logs with any ElasticSearch framework you have to first get access to them... and to do so, I was looking to the "gotk-components.yaml" manifest that is created with Flux bootstrap but didn't find in which volume the logs are created or in which folder inside the volume.

Any hint you can provide on where to find them would be great. Also, do you suggest using a persistent volume to store the logs? If so, should I adjust the above manifest manually after its creation and "reconcile"? -> `flux reconcile kustomization flux-system --with-source`

Thanks again!

**Reply by stefanprodan** [reply]
Kubernetes controllers don’t store the logs in volumes, this is managed by the CRI. As suggested above you need to setup log aggregation.

**Reply by stefanprodan** [reply]
Here is one way to do it https://github.com/fluent/fluent-bit-kubernetes-logging

**Comment by matmisie-gemini** [comment]
Maybe I misunderstood you but it looks like you just want to print recent error messages and not get the whole container logs (btw this is trimmed to 10mb by default by k8s). Try passing --tail 100 to kubectl logs?

**Reply by stefanprodan** [reply]
There is `flux logs --tail=100 --follow --level=error --all-namespaces` that streams the logs from all the controllers, docs here: https://toolkit.fluxcd.io/cmd/flux_logs/
